client/extensions/woocommerce/state/settings/shipping/zones/reducer.js

```diff
@@ -0,0 +1,179 @@
+	return { ...state,
+		currentlyEditingZone: {
+			id: null,
+			locations: [],
+			methods: [],
+			order: ( ( state.zones[ state.zones.length - 2 ] || {} ).order || 0 ) + 1,
+		},
+		currentlyEditingZoneIndex: -1,
+	};
+};
+
+reducers[ WOOCOMMERCE_SHIPPING_ZONE_EDIT ] = ( state, { index } ) => {
+	return { ...state,
+		currentlyEditingZone: state.zones[ index ],
+		currentlyEditingZoneIndex: index,
+	};
+};
+
+reducers[ WOOCOMMERCE_SHIPPING_ZONE_REMOVE ] = ( state, { index } ) => {
+	// Protect { id: 0 } (Rest of the World) from deletion
+	if ( 0 === state.zones[ index ].id ) {
+		return state;
+	}
+	return { ...state,
+		currentlyEditingZone: null,
+		zones: [ ...state.zones.slice( 0, index ), ...state.zones.slice( index + 1 ) ],
+	};
+};
+
+reducers[ WOOCOMMERCE_SHIPPING_ZONE_CANCEL ] = ( state ) => {
+	return { ...state,
+		currentlyEditingZone: null,
+	};
+};
+
+reducers[ WOOCOMMERCE_SHIPPING_ZONE_CLOSE ] = ( state ) => {
+	if ( ! state.currentlyEditingZone ) {
+		return state;
+	}
+	const zones = [ ...state.zones ];
+	if ( -1 === state.currentlyEditingZoneIndex ) {
+		// Keep "Rest Of The World" last always
+		zones.splice( zones.length - 1, 0, state.currentlyEditingZone );
+	} else {
+		zones[ state.currentlyEditingZoneIndex ] = state.currentlyEditingZone;
+	}
+	return { ...state,
+		zones,
+		currentlyEditingZone: null,
+	};
+};
+
+reducers[ WOOCOMMERCE_SHIPPING_ZONE_LOCATION_ADD ] = ( state, { locationType, locationCode } ) => {
+	// TODO: ZIP codes
+	// TODO: Get the list of continents / countries / states from somewhere (new endpoint?)
+	if ( ! state.currentlyEditingZone ||
+		0 === state.currentlyEditingZone.id ||
+		! LOCATION_TYPES.includes( locationType ) ) {
+		return state;
+	}
+	const location = { type: locationType, code: locationCode };
+	const newLocations = unionWith( state.currentlyEditingZone.locations, [ location ], isEqual );
+	return { ...state,
+		currentlyEditingZone: { ...state.currentlyEditingZone,
+			locations: newLocations,
+		},
+	};
+};
+
+reducers[ WOOCOMMERCE_SHIPPING_ZONE_LOCATION_REMOVE ] = ( state, { locationType, locationCode } ) => {
+	// TODO: when removing a country, remove all its states?
+	if ( ! state.currentlyEditingZone || ! LOCATION_TYPES.includes( locationType ) ) {
+		return state;
+	}
+	const location = { type: locationType, code: locationCode };
+	const newLocations = differenceWith( state.currentlyEditingZone.locations, [ location ], isEqual );
+	return { ...state,
+		currentlyEditingZone: { ...state.currentlyEditingZone,
+			locations: newLocations,
+		},
+	};
+};
+
+reducers[ WOOCOMMERCE_SHIPPING_ZONE_METHOD_ADD ] = ( state ) => {
+	if ( isEmpty( state.methodDefinitions ) || ! state.currentlyEditingZone ) {
+		return state;
+	}
+	const currentMethods = state.currentlyEditingZone.methods;
+	const methodDefinition = {
+		method_id: state.methodDefinitions[ 0 ].id,
+		id: null,
+		order: ( ( currentMethods[ currentMethods.length - 1 ] || {} ).order || 0 ) + 1,
```

Comment:
The UI only really cares about `index`, but `order` is needed for actually making the `REST` create request. The problem is that `order` is `1`-based, and it isn't necessarily sequiential (you can have 3 zones, with `order` `1`, `7` and `42`).